src/webgl/program.js

```diff
@@ -50,10 +50,18 @@ export default class Program extends Resource {
     }
   }
 
-  initialize({vs, fs, defaultUniforms, varyings, bufferMode = GL.SEPARATE_ATTRIBS} = {}) {
+  initialize({
+    vs,
+    fs,
+    defaultUniforms,
+    varyings,
+    bufferMode = GL.SEPARATE_ATTRIBS,
+    cacheKey = null
```

Comment:
Nit: Do we need to store it?

Reply:
Yes, `Program`, doesn't store shader strings, it only stores `VertexShader`/`FragmentShader` objects.

Reply:
True, my thinking was that the shader objects do store the source code.

Reply:
Yeah, I can access using `source`, but also we should 'key' calculation in one place, in this case during `getProgram` and store the same key inside the `Program`?

Reply:
> also we should keep 'key' calculation in one place

Sure, no worries, just to explain my thinking:
* One of my favorite principles is separation of concerns, i.e. to keep the implementation details of each subsystem (class) inside that system and expose minimal interfaces between them
* Accordingly, an ideal design in my mind is one where `Program` does not even know of the existence of the `ShaderCache` which is a "higher level" class.
* However, the cache key is very much an implementation detail of the `ShaderCache` - which currently chooses to use long strings as keys but that could internally be changed to e.g. integers, or double strings, or shader handles etc without modifying any other classes).
* By putting the code that creates that key in the `Program` we are creating a pretty strong coupling between Program and ShaderCache (in the sense that ShaderCache relies on program to know how to calculate keys that fit the internal structure of `ShaderCache`).

Since the `source` is a public field of the `Shader` class we could just do:
```js
ShaderCache {
  _getProgramKey(program) {
     return `${program.vs.source}-${program.fs.source}
  }
}
```
Which would improve the separation of concerns, in my view.